source/language/openpulse.rst

```diff
@@ -0,0 +1,742 @@
+
+        // Loop over delay times from ``0`` to ``T``
+        length T = ...;
+        int[32] steps = ...;
+        length inc_time = T/steps;
+        for τ in [0:inc_time:T]:
+            // Simultaneous π/2 pulses
+            transmit(eom_a, raman_a, pi2_time);
+            transmit(eom_b, raman_b, pi2_time);
+            transmit(aod, sum(q1_pi2_sig, q2_pi2_sig, q3_pi2_sig), pi2_time);
+
+            // Delay each channel
+            delay(eom_a, eom_b, aod, τ/2);
+
+            // π pulse on qubit 1 only -- composed of two π/2 pulses (for Hahn echo)
+            for i in [0:2] {
+                transmit(eom_a, raman_a, pi2_time);
+                transmit(eom_b, raman_b, pi2_time);
+                transmit(aod, q1_pi2_sig, pi2_time);
+            }
+
+            // Delay each channel again
+            delay(eom_a, eom_b, aod, τ/2);
+
+            // Phase shift the signals by diff amounts
+            float[64] tppi_1 = ...;
+            float[64] tppi_2 = ...;
+            float[64] tppi_2 = ...;
+            q1_carrier.phase += tppi_1 * τ;  // autocast to angle[size]
+            q2_carrier.phase += tppi_2 * τ;
+            q3_carrier.phase += tppi_3 * τ;
+
+            // Simultaneous π/2 pulses
+            transmit(eom_a, raman_a, pi2_time);
+            transmit(eom_b, raman_b, pi2_time);
+            transmit(aod, sum(q1_pi2_sig, q2_pi2_sig, q3_pi2_sig), pi2_time);
+    }
+
+Measurement
+~~~~~~~~~~~
+
+Demonstrates how to define a measurement of a superconducting qubit through a dispersive readout
+process within the signal framework.
+
+.. code-block:: c
+
+    // Use a boxcar function to generate IQ data from raw signal
+    kernel boxcar(signal input)->complex[64];
+    // Use a linear discriminator to generate bits from IQ data
+    kernel discriminate(complex[64] iq)->bit;
+
+    defcal measure $0 -> bit {
+        // Define the channels
+        txchannel m0 = txch($0, "measure");
+        rxchannel cap0 = rxch($0, "capture");
+
+        // Force time of carrier to 0 for consistent phase for discrimination.
+        signal carr = set(exp(1.0, 5e9, 0), 0);
+
+        // Apply measurement stimulus
+        envelope meas_pulse = gaussian_square(1.0, 16000dt, 262dt, 13952dt);
+
+        // Transmit signal
+        transmit(m0, mix(carrier, meas_pulse), meas_pulse.duration);
+        // Align measure and capture channels
+        barrier(m0, cap0);
+        // Capture transmitted signal after interaction with measurement resonator
+        signal raw_output = receive(cap0, meas_pulse.duration);
+
+        // Kernel and discriminate
+        complex[32] iq = boxcar(raw_output);
+        bit result = discriminate(iq);
+
+        return result;
+    }
+
+Rz gate (Virtual-Z Gate)
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+Demonstrates how to implement an Rz gate (e.g. Virtual-Z gate) with signals.
+
+.. code-block:: c
+
+    defcal rz(angle[20] theta) $q {
+        txchannel dq = txch("drive", $q);
```

Comment:
Do I need the `txchannel` here?

Reply:
Yes. This is the issue w/ not having globals, ie you have to redefine the channel every time. We can work out a preferred way in the frame proposal.